cmd/ctr/commands/run/run.go

```diff
@@ -18,19 +20,128 @@ import (
+type containerConfig struct {
+	Checkpoint  string   `toml:"checkpoint"`
+	Cwd         string   `toml:"cwd"`
+	Env         []string `toml:"env"`
+	Label       []string `toml:"label"`
+	Mount       []string `toml:"mount"`
+	NetHost     bool     `toml:"net_host"`
+	ReadOnly    bool     `toml:"read_only"`
+	Runtime     string   `toml:"runtime"`
+	Snapshotter string   `toml:"snapshotter"`
+	TTY         bool     `toml:"tty"`
+	// Unix-specific field
+	RootFS bool `toml:"rootfs"`
+	// Windows-specific field
+	Layer []string `toml:"layers"`
+}
+
+// ContainerFlags are cli flags specifying container options
+var ContainerFlags = []cli.Flag{
+	cli.StringFlag{
+		Name:  "config,c",
+		Usage: "path to the configuration file",
+	},
+	cli.StringFlag{
+		Name:  "checkpoint",
+		Usage: "provide the checkpoint digest to restore the container",
+	},
+	cli.StringFlag{
+		Name:  "cwd",
+		Usage: "specify the working directory of the process",
+	},
+	cli.StringSliceFlag{
+		Name:  "env",
+		Usage: "specify additional container environment variables (i.e. FOO=bar)",
+	},
+	cli.StringSliceFlag{
+		Name:  "label",
+		Usage: "specify additional labels (i.e. foo=bar)",
+	},
+	cli.StringSliceFlag{
+		Name:  "mount",
+		Usage: "specify additional container mount (ex: type=bind,src=/tmp,dest=/host,options=rbind:ro)",
+	},
+	cli.BoolFlag{
+		Name:  "net-host",
+		Usage: "enable host networking for the container",
+	},
+	cli.BoolFlag{
+		Name:  "read-only",
+		Usage: "set the containers filesystem as readonly",
+	},
+	cli.StringFlag{
+		Name:  "runtime",
+		Usage: "runtime name (io.containerd.runtime.v1.linux, io.containerd.runtime.v1.windows, io.containerd.runtime.v1.com.vmware.linux)",
+		Value: fmt.Sprintf("io.containerd.runtime.v1.%s", runtime.GOOS),
+	},
+	cli.BoolFlag{
+		Name:  "tty,t",
+		Usage: "allocate a TTY for the container",
+	},
+}
+
+func loadContainerConfig(path string, c *containerConfig) error {
+	if _, err := toml.DecodeFile(path, c); err != nil {
+		if os.IsNotExist(err) {
+			return errors.New("please specify the config file")
+		}
+		return err
+	}
+	return nil
+}
+
+func applyFlags(context *cli.Context, config *containerConfig) {
+	if cp := context.String("checkpoint"); cp != "" {
+		config.Checkpoint = cp
+	}
+	if c := context.String("cwd"); c != "" {
+		config.Cwd = c
+	}
+	if e := context.StringSlice("env"); len(e) > 0 {
+		config.Env = e
+	}
+	if l := context.StringSlice("label"); len(l) > 0 {
+		config.Label = l
+	}
+	if m := context.StringSlice("mount"); len(m) > 0 {
+		config.Mount = m
+	}
+	if n := context.Bool("net-host"); n {
+		config.NetHost = n
+	}
+	if ro := context.Bool("read-only"); ro {
+		config.ReadOnly = ro
+	}
+	if r := context.String("runtime"); r != "" {
+		config.Runtime = r
+	}
+	if t := context.Bool("tty"); t {
+		config.TTY = t
+	}
+	// if runtime field is empty, flag will set value
+	if config.Runtime == "" {
+		config.Runtime = context.String("runtime")
+	}
+	// if snapshotter field is empty, flag will set value
```

Comment:
If I understand this correctly the flags will override the config in most cases, but in these 2 cases the config has higher precedence. Why is that?

Reply:
these two flags will set default values if no flag has been set (see above). since they will never be empty a "" check will always override what's in the config file even if the user never set a flag. i realize this behavior is different so i'm open to suggestions on how to handle them!

Reply:
Ah! I'm not familiar with urfave/cli, but I think you can use https://godoc.org/github.com/urfave/cli#Context.IsSet to figure out if the flag was set by the user. That way the flag can still override the config file.

Reply:
lol that would be dope! looking into that! thanks :)